crates/services/txpool/src/service.rs

```diff
@@ -313,13 +332,34 @@ where
     DB: TxPoolDb,
 {
     #[tracing::instrument(name = "insert_submitted_txn", skip_all)]
-    pub fn insert(
+    pub async fn insert(
         &self,
         txs: Vec<Arc<Transaction>>,
     ) -> Vec<anyhow::Result<InsertionResult>> {
-        let insert = { self.txpool.lock().insert(&self.tx_status_sender, &txs) };
+        // verify txs
+        let block_height = self.db.current_block_height();
+        let current_height = match block_height {
+            Ok(val) => val,
+            Err(e) => return vec![Err(e.into())],
+        };
+
+        let checked_txs = check_transactions(&txs, current_height, &self.config).await;
+
+        let valid_txs = checked_txs
+            .iter()
+            .filter_map(|tx_chech| match tx_chech {
+                Ok(tx) => Some(tx.clone()),
+                Err(e) => {
+                    tracing::error!("Transaction is invalid: {}", e);
+                    None
+                }
+            })
+            .collect();
```

Comment:
It would be nice if you could do that without cloning of the transaction. For example, you can wrap all values into `Some(result)`. After, you can extract all valid transactions and replace them with `None`, insert them, and return back. It is one available option, but maybe you can find another one=D